### ISSUE TITLE
 Possible race in wayland_sync 

### ISSUE THREAD

**Issue by phako** [01org/gstreamer-media-SDK#2]
Text copied from upstream vaapi bug, code is the same here.

If you have a second thread that registers a second queue on the wayland display and that will call wl_display_flush, it can happen that the frame callbacks are already triggered by the while loop in https://cgit.freedesktop.org/gstreamer/gstreamer-vaapi/tree/gst-libs/gst/vaapi/gstvaapiwindow_wayland.c#n174. So it can happen that after 

https://cgit.freedesktop.org/gstreamer/gstreamer-vaapi/tree/gst-libs/gst/vaapi/gstvaapiwindow_wayland.c#n178

the priv->num_frames_pending is already 0 and we will be stuck in gst_poll_wait forever.

**Comment by raj23** [comment]
Hi,

Thanks for pointing this out to us. 
We would like to understand under what circumstances this will happen? 

Thank you.

**Comment by phako** [comment]
Say you have the video running full screen and need touch events from that surface. 
We get the display from the sink and register an event queue for wayland input events there and handle that in a separate thread (prepare_read etc.)

Due to parallelism, it can happen that the input thread calls wl_display_flush before the wait function clears its event queue, so attach/damage/commit will be sent to the server and the frame event will already be triggered by the queue clearing.

We have seen this with vaapisink but since your code is derived from that (and we will probably have to switch to it at some point) it should behave the same in that situation

Note, this is theoratical as we have seen it with vaapisink but we cannot run mfx stuff yet.

Btw, how are you handling generic bugfixes that happen(ed) to vaapi? Do you regular backsyncs of generic handling?

**Comment by raj23** [comment]
Thank you for the explanation. Yes I believe it will have the same behavior as vaapisink as the code is derived from that. 

We don't handle the general bugfixes that happened to vaapi and we don't regularly backsync of generic handling. When we started this, we derived from gstreamer-vaapi but now our development to some extent is detached with gstreamer-vaapi and currently we in the mid of refactoring our codes. We couldn't make it public yet as the codebase is not stable.
